src/starks/verifier.rs

```diff
@@ -265,31 +264,47 @@ fn step_2_verify_claimed_composition_polynomial<F: IsFFTField, A: AIR<Field = F>
         &challenges.rap_challenges,
     );
 
-    let divisor_x_n = (&challenges.z.pow(trace_length) - FieldElement::<F>::one()).inv();
+    let denominator = (&challenges.z.pow(trace_length) - FieldElement::<F>::one()).inv();
 
-    let denominators = air
-        .transition_exemptions_verifier()
+    let exemption = air
+        .transition_exemptions_verifier(
+            domain.trace_roots_of_unity.iter().last().expect("has last"),
+        )
         .iter()
-        .map(|poly| poly.evaluate(&challenges.z) * &divisor_x_n)
+        .map(|poly| poly.evaluate(&challenges.z))
         .collect::<Vec<FieldElement<F>>>();
 
-    let degree_adjustments = air
+    let max_degree = air
         .context()
         .transition_degrees()
         .iter()
+        .max()
+        .expect("has maximum degree");
+    let degree_adjustments = (1..=*max_degree)
         .map(|transition_degree| {
             let degree_adjustment =
                 composition_poly_degree_bound - (trace_length * (transition_degree - 1));
             challenges.z.pow(degree_adjustment)
         })
         .collect::<Vec<FieldElement<F>>>();
 
-    let transition_c_i_evaluations_sum =
-        ConstraintEvaluator::<F, A>::compute_constraint_composition_poly_evaluations_sum(
-            &transition_ood_frame_evaluations,
-            &denominators,
-            &degree_adjustments,
-            &challenges.transition_coeffs,
+    let transition_c_i_evaluations_sum = transition_ood_frame_evaluations
+        .iter()
+        .zip(&air.context().transition_degrees)
+        .zip(&air.context().transition_exemptions)
+        .zip(&challenges.transition_coeffs)
+        .fold(
+            FieldElement::zero(),
+            |acc, (((eval, degree), except), (alpha, beta))| {
+                if except == &0 {
+                    acc + &denominator * eval * (alpha * &degree_adjustments[degree - 1] + beta)
+                } else {
+                    acc + &denominator
+                        * eval
+                        * (alpha * &degree_adjustments[degree - 1] + beta)
+                        * &exemption[except - 1]
+                }
```

Comment:
I believe this would be easier to read this way:
```suggestion
                let except = except.checked_sub(1).map(|i| &exemption[i]).unwrap_or(&1);
                acc + &denominator * eval * (alpha * &degree_adjustments[degree - 1] + beta) * except
```
This way there's much fewer logic inside the different branches.